book/numerical_methods/curve_fitting.nim

```diff
@@ -29,6 +31,253 @@ yes.
+    fitFunc(solution, x)
+
+  var df = toDf({"t": tDense, "original": yDense, "fitted": ySol, "tPoint": t, "yPoint": y})
+  df = df.gather(@["original", "fitted"], key="Class", value = "y")
+  df.ggplot(aes("t", "y", color="Class")) +
+    geom_line() +
+    geom_point(aes("tPoint", "yPoint")) +
+    ggsave("images/levmarq_comparision.png")
+
+nbImage("images/levmarq_comparision.png")
+
+nbText: hlMd"""
+As we can see, the fitted curve is quite close to the original one.
+
+## Errors & Uncertainties
+We might also want to quantify exactly how good of a fit the computed weights give.
+One measure commonly used is [$\chi^2$](https://en.m.wikipedia.org/wiki/Pearson%27s_chi-squared_test):
+$$\chi^2 = \sum_i \frac{(y_i - \hat{y}(x_i))^2}{\sigma_i^2}$$ 
+where $y_i$ and $x_i$ are the measurements, $\hat{y}$ is the fitted curve and $\sigma_i$
+is the standard deviation (uncertainty/error) of each of the measurements. We will use the
+noise size we used when generating the samples here, but in general you will have to
+figure out yourself what errors are in your situation. It may for example be the resolution
+of your measurements or you may have to approximate it using the data itself. 
+
+We now create the error vector and sample the fitted curve:
+"""
+
+nbCode:
+  let yError = ones_like(y) * noise
+  let yCurve = t.map_inline:
+    fitFunc(solution, x)
+
+nbText: """
+Now we can calculate the $\chi^2$:
+"""
+
+nbCode:
+  var chi2 = 0.0
+  for i in 0 ..< y.len:
+    chi2 += ((y[i] - yCurve[i]) / yError[i]) ^ 2
+  echo "χ² = ", chi2
+
+nbText: hlMd"""
+Great! Now we have a measure of how good the fit is, but what if we add more points?
+Then we will get a better fit, but we will also get more points to sum over.
+And what if we choose another curve to fit with more parameters?
+Then we may be able to get a better fit but we risk overfitting.
+[Reduced $\chi^2$](https://en.m.wikipedia.org/wiki/Reduced_chi-squared_statistic)
+is a measure which adjusts $\chi^2$ to take these factors into account.
+The formula is:
+$$\chi^2_{\nu} = \frac{\chi^2}{n_{\text{obs}} - m_{\text{params}}}$$
+where $n_{\text{obs}}$ is the number of observations and $m_{\text{params}}$
+is the number of parameters in the curve. The difference between them is denote
+the degree of freedom (dof).
+This is simplified, the mean $\chi^2$
+score adjusted to penalize too complex curves.
+
+Let's calculate it!
+"""
+
+nbCode:
+  let reducedChi2 = chi2 / (y.len - solution.len).float
+  echo "Reduced χ² = ", reducedChi2
+
+nbText: hlMd"""
+As a rule of thumb, values around 1 are desirable. If it is much larger
+than 1, it indicates a bad fit. And if it is much smaller than 1 it means
+that the fit is much better than the uncertainties suggested. This could
+either mean that it has overfitted or that the errors were overestimated.
+In our example I suspect the later is the case as the actual standard
+deviation of the error is smaller than the maximum value that we used here.
```

Comment:
The reason for the deviation from 1 is very likely that the noise we use is uniform. However, the χ² test essentially assumes the given weight (or error) to correspond to a 1σ environment of a _normal distributed_ uncertainty. If you were to "map" the error of the `noise` from uniform to a normal distribution and expect the to hand a 1σ environment, that should imply to use an error of more or less `noise * 0.66`. Doing that should give you a number close to 1. 

As the context of the tutorial is not to be about goodness of fit tests etc. your explanation is fine though. Just as an FYI if you wondered why you're seeing what you're seeing.

Reply:
Indeed :+1: 
I finally found the `randomNormalTensor` function now and using it gives a χ²/dof of 1.06 :smile: So I'll remove the last sentence.